pkgs/build-support/checkpoint-build.nix

```diff
@@ -44,26 +48,31 @@ rec {
   });
 
   /* Build a derivation based on the checkpoint output generated by
-    * the `prepareCheckpointBuild function.
+    * the `prepareCheckpointBuild` function.
     *
     * Usage:
     * let
-    *   checkpointArtifacts = prepareCheckpointBuild drv
-    * in mkCheckpointedBuild drv checkpointArtifacts
+    *   checkpointArtifacts = prepareCheckpointBuild drv;
+    * in mkCheckpointBuild drv checkpointArtifacts
   */
-  mkCheckpointedBuild = drv: previousBuildArtifacts: drv.overrideAttrs (old: {
+  mkCheckpointBuild = drv: checkpointArtifacts: drv.overrideAttrs (old: {
     # The actual checkpoint build phase.
-    # We compare the changed sources from a previous build with the current and create a patch
-    # Afterwards we clean the build directory to copy the previous output files (Including the sources)
-    # The source difference patch is applied to get the latest changes again to allow short build times.
+    # We compare the changed sources from a previous build with the current and create a patch.
+    # Afterwards we clean the build directory and copy the previous output files (including the sources).
+    # The source difference patch is then applied to get the latest changes again to allow short build times.
     preBuild = (old.preBuild or "") + ''
       set +e
-      diff -ur ${previousBuildArtifacts}/sources ./ > sourceDifference.patch
+      sourceDifferencePatchFile=$(${mktemp}/bin/mktemp)
+      diff -ur ${checkpointArtifacts}/sources ./ > "$sourceDifferencePatchFile"
       set -e
-      shopt -s extglob dotglob
-      rm -r !("sourceDifference.patch")
-      ${pkgs.rsync}/bin/rsync -cutU --chown=$USER:$USER --chmod=+w -r ${previousBuildArtifacts}/outputs/* .
-      patch -p 1 -i sourceDifference.patch
+      rm -r *
+      ${rsync}/bin/rsync -cutU --chown=$USER:$USER --chmod=+w -r ${checkpointArtifacts}/outputs/* .
+      patch -p 1 -i "$sourceDifferencePatchFile"
+      rm "$sourceDifferencePatchFile"
     '';
   });
+
+  mkCheckpointedBuild = lib.warn
+    "`mkCheckpointedBuild` is deprecated, use `mkCheckpointBuild` instead!"
+    mkCheckpointBuild;
```

Comment:
Considering how recent this is, I think we can backport this PR even without removing the warning.

Reply:
> without removing the warning

Do you mean `without the warning`? :laughing: To make sure that I understand this correctly:
- we keep the warning as-is in this PR,
- but remove it in the backport-23.11 PR
Or, should we just remove it altogether? Note that [currently there is usage of it in nixpkgs](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+mkCheckpointedBuild&type=code).

Reply:
I have **removed the old `mkCheckpointedBuild = lib.warn ...` completely**. Please ping me if this is not desired!

Reply:
I'd prefer a deprecation warning still. Some people have already been linked to the original PR and probably tried it out. And it doesn't hurt to have such a warning :)

Reply:
Restored!